Enable compiler warnings

wjt opened the pull request:
This covers `gnome-image-installer/`. Some code from `gnome-initial-setup/` is used but I'd rather fix those in upstream g-i-s and pull the changes back down. In a few places there were bugs there exposed by warnings in the installer-specific code, so I've fixed those as needed.

https://phabricator.endlessm.com/T15159

wjt committed:
gis-page: constify page_id

Every implementation assigns a const string to it, and no non-const
operations are performed on it.

pwithnall commented on configure.ac:
```diff
@@ -132,8 +132,17 @@ fi
-GNOME_MAINTAINER_MODE_DEFINES
+# -Wsign-compare:
+#   I was motivated to wire up warnings by hours wasted on checking if a
+#   unsigned value is negative, so -Wsign-compare is non-negotiable.
```
If `-Wsign-compare` is missing from the default upstream `AX_COMPILER_FLAGS` warning set and you believe it has universal value, please submit a patch [upstream](https://github.com/peti/autoconf-archive) and CC me.

wjt replied:
Actually it's already there; I must have added this before switching from the GNOME one, maybe. I'll remove it.

pwithnall commented on gnome-image-installer/gnome-image-installer.c:
```diff
@@ -365,14 +361,6 @@ gis_page_get_type();
-    g_critical ("Clutter-GTK init failed");
-    exit (1);
-  }
-#endif
```
Why remove these? Deprecated? Unused? Could do with a comment in the commit message.

wjt replied:
Unused, inherited from gnome-initial-setup's `main()` I assume. I'll amend.

pwithnall commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -103,7 +103,7 @@ gis_diskimage_page_selection_changed(GtkWidget *combo, GisPage *page)
     {
-      gint64 size = get_gzip_disk_image_size (image);
+      guint64 size = get_gzip_disk_image_size (image);
       if (size <= 0)
```
These `size <= 0` comparisons can be clarified as `size == 0` now (in several places below as well).

wjt replied:
Good catch.

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -619,7 +617,7 @@ gis_install_page_verify (GisPage *page)
   GFile *signature = g_file_new_for_path (signature_path);
   gint outfd;
-  gchar *args[] = { "gpg",
+  const gchar *args[] = { "gpg",
```
This could actually be `const gchar * const args[]`, I think? Because both the array and each value are `const`.

Disclaimer: I haven’t tried this, and my knowledge of C array-vs-`const` semantics is poor at best.

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -637,7 +635,7 @@ gis_install_page_verify (GisPage *page)
     }
 
-  if (!g_spawn_async_with_pipes (NULL, args, NULL,
+  if (!g_spawn_async_with_pipes (NULL, (gchar **) args, NULL,
```
IIRC this cast is the best way to tackle this problem, from the times I’ve tried to come up with a better solution in the past.

pwithnall reviewed (changes requested):
As a general comment, I would have done the warning fixes first as separate commits per class of warning; then enabled the compiler warnings as a final commit. Allows for separate review and reversion of related groups of things.

However, that’s just a nitpick and I’m not asking you to rearrange all these commits.

wjt committed:
main: remove unused init inherited from g-i-s

Actually, g-i-s hasn't used clutter or cheese for a long time, either.

Our fork of g-i-s does use evince, but eos-installer doesn't.

We use a different set of pages to g-i-s.

wjt committed:
Add missing #includes

wjt committed:
Add missing (void) in function prototypes

wjt committed:
Move GQuark function definitions to headers

wjt committed:
Consistently use uint64 for installed size of images

Disk sizes are by definition strictly positive, and indeed
get_gzip_disk_image_size () and friends return unsigned values, so the
store should store an unsigned value too. This fixes a
signed-vs-unsigned comparison warning in disktarget, which compares the
uncompressed image size to the candidate target device size (which is
unsigned).

While we're here, simplify some dodgy logic dealing with the (supposedly
impossible) case where we can't determine the uncompressed size of the
image. Rather than lazily determining the uncompressed size, we store it
in the list model along with all the other facts.

wjt committed:
Remove unused locals

wjt committed:
"Fix" const-ness warnings when running gpg

execve's argv argument is const-ish but for some reason Gio's wrapper is
not. In practice everyone casts the const away.

wjt committed:
gpt_lzma: declare private function as 'static'

wjt committed:
gis_page_set_title(): constify argument

The string is dup'd anyway.

wjt committed:
Enable many fatal warnings

https://phabricator.endlessm.com/T15159

pwithnall commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -103,34 +104,26 @@ gis_diskimage_page_selection_changed(GtkWidget *combo, GisPage *page)
+  if (required_size == 0)
+    {
+      g_warn_if_fail (required_size > 0);
+      required_size = G_MAXINT64;
```
This will now warn if the `image` name doesn’t match any of the suffix checks above. Are you confident that won’t be the case? (If so, maybe add an `else { g_assert_not_reached (); }`?)

pwithnall commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -70,7 +70,8 @@ enum {
     IMAGE_SIGNATURE,
-    ALIGN
+    ALIGN,
+    IMAGE_REQUIRED_SIZE
```
Nitpick: I’d add a trailing comma here to reduce diff noise in future. (But there’s no need to fix this now; just for future pondering.)

pwithnall commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -250,11 +224,31 @@ add_image (
-       || (g_str_has_suffix (image, ".img.xz") && get_xz_is_valid_eos_gpt (image) == 1)
-       || (g_str_has_suffix (image, ".img") && get_is_valid_eos_gpt (image) == 1)
-       || (image_device != NULL && get_is_valid_eos_gpt (image_device) == 1))
+      if (required_size != 0)
```
Nice!

pwithnall reviewed (approved):
Nice one. I’ll merge shortly.
